8357600: Patch nmethod flushing message to include more details

JohnTortugo opened the pull request:
Please review this patch for adding more details to nmethod flushing message. These details are particularly important when investigating interaction of JVMCI compiled code and code cache flushing heuristics.

Tested on Linux x64 with JTREG tier1-3 using fastdebug and release builds.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8357600](https://bugs.openjdk.org/browse/JDK-8357600): Patch nmethod flushing message to include more details (**Enhancement** - P4)


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25402/head:pull/25402` \
`$ git checkout pull/25402`

Update a local copy of the PR: \
`$ git checkout pull/25402` \
`$ git pull https://git.openjdk.org/jdk.git pull/25402/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25402`

View PR using the GUI difftool: \
`$ git pr show -t 25402`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25402.diff">https://git.openjdk.org/jdk/pull/25402.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25402#issuecomment-2902773164)
</details>

JohnTortugo committed:
Add more info to nmethod flushing.

shipilev reviewed (commented):
A bit concerned about performance impact of this logging, especially since we are under `CodeCache_lock`. So I would suggest two improvements:

1. Maybe move logging before acquiring `CodeCache_lock`? Not sure if it is safe for various `CodeCache::*` getters.

2. Predicate the argument preparation/logging with:

``` 
  LogTarget(Debug, codecache) lt;
  if (lt.is_enabled()) {
    ...
```

JohnTortugo committed:
Move before lock & check logging

JohnTortugo commented:
@shipilev - I made changes following your suggestion. I moved the printing to before acquiring the lock and I also moved it to under a if checking if logging is enabled. I don't think these get/logging operation in particular need to be under a lock.

shipilev reviewed (commented):
Not sure about calling `is_cold()` without a lock: looks like `_gc_epoch` counter is unsynchronized. Now that you predicated the bulk of the logging with logging checks, I think it is fine to grab `CodeCache_lock` before doing logging.

JohnTortugo committed:
Address PR feedback.

shipilev reviewed (approved):
Looks OK to me. This is a diagnostic logging, so we do not have to be extra crisp about it. Let any other compiler folks review as well.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2123,10 +2123,27 @@ void nmethod::purge(bool unregister_nmethod) {
+      is_jvmci = "JVMCI compiled ";
+    }
+#endif
+    log_debug(codecache)("Flushing nmethod %3d/" INTPTR_FORMAT ", level=%d, osr=%d, cold=%d, epoch=" UINT64_FORMAT ", cold_count=" UINT64_FORMAT ". "
```
You can use `lt` here.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2123,10 +2123,27 @@ void nmethod::purge(bool unregister_nmethod) {
+    log_debug(codecache)("Flushing nmethod %3d/" INTPTR_FORMAT ", level=%d, osr=%d, cold=%d, epoch=" UINT64_FORMAT ", cold_count=" UINT64_FORMAT ". "
+                       "Cache capacity: %zuKb, free space: %zuKb. %smethod %s",
+                       _compile_id, p2i(this), _comp_level, is_osr_method(), is_cold(), _gc_epoch, CodeCache::cold_gc_count(),
+                       codecache_capacity, codecache_free_space, is_jvmci, name);
```
Swap `is_jvmci` and `name` so that method's names in output for nmethod not compiled by JVMCI (by C1) will be aligned.

vnkozlov commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -2123,10 +2123,27 @@ void nmethod::purge(bool unregister_nmethod) {
+  if (lt.is_enabled()) {
+    ResourceMark rm;
+    const char* name = method()->name()->as_C_string();
+    const char* is_jvmci = "";
```
Please use `compiler_name()` instead.

JohnTortugo committed:
Refactoring: use compiler_name(), LogStream

vnkozlov reviewed (approved):
Good.

JohnTortugo commented:
/integrate
